src/navmesh.js

```diff
@@ -373,98 +375,102 @@ export class NavMesh {
-
-            // Calculate portal edges
-            let [newEdge1, newEdge2] = this._pointPortalEdges(from, portal);
-
-            if (edge1 !== null && edge2 !== null) {
-                // Shrink funnel edges
-                newEdge1 = this._funnelEdge(edge1, edge2, newEdge1);
-                newEdge2 = this._funnelEdge(edge2, edge1, newEdge2);
-
-                if (newEdge1 === null || newEdge2 === null) {
-                    const edge = newEdge1 === null ? edge2 : edge1;
-                    const newPath = this._splitAt(
-                        // We only need to check the polygon up to now, not the future
-                        // ones. This can have serious performance implications,
-                        // as the function is recursive.
-                        path.slice(i - 1),
-                        edge.p2
-                    );
-                    points.push(...this._funnel(edge.p2, to, newPath));
-                    break;
-                }
-            }
+            const nextPoly = path[i + 1];
+            const portal = poly.neighbors[nextPoly._uuid].portal;
+            // The portal end points are in clockwise order, viewed from the inside of the polygon.
 
-            edge1 = newEdge1;
-            edge2 = newEdge2;
-
-            // If we are at the end of the path,
-            // just add the last edge, which is from -> to
-            if (i === path.length - 2) {
-                const toEdge = new Edge(from, to);
-                let newToEdge = this._funnelEdge(edge1, edge2, toEdge);
-                if (newToEdge === edge1) {
-                    points.push(edge1.p2);
-                } else if (newToEdge === null) {
-                    points.push(edge2.p2);
-                }
+            // Extend funnel on the left
+            this._extendFunnel(tail, left, right, true, portal.p1);
+
+            // Extend funnel on the right
+            this._extendFunnel(tail, left, right, false, portal.p2);
+        }
+
+        // Close funnel to endpoint
+        this._extendFunnel(tail, left, right, true, to);
+        this._extendFunnel(tail, left, right, false, to);
 
-                points.push(to);
+        return tail;
+    }
+
+    _extendFunnel(tail, left, right, extendLeft, newPoint){
+        const apex = tail[tail.length - 1];
+        // We pretend to be in the `expandLeft` case here. Otherwise flip.
+        if (!extendLeft) {
+            [left, right] = [right, left];
+        }
+
+        // If `newPoint` is the end point of the left side of the funnel, skip it.
+        const lastLeft = left.length === 0 ? tail[tail.length - 1] : left[left.length - 1];
+        if (newPoint.equals(lastLeft)) {
+            return;
+        }
+
+        // Determine angle of `apex`-`newPoint` relative to `apex`-`left[j]`
+        let j = 0;
+        while (j < left.length
+                && this._isInClockwiseOrder(apex, newPoint, left[j], !extendLeft)) {
+            j++;
+        }
```

Comment:
We're also doing this looping twice, so it would be nice if we could encapsulate this operation into a helper method.

Another point is that I would maybe find clearer using a `for ... of ...` loop instead of a `while` loop, but that's up to you.

Reply:
I converted into a `for`-loop. A `for ... of`-loop isn't possible I think because I need the index.